tests/jobserver/test_releases.py

```diff
@@ -0,0 +1,106 @@
+from zipfile import ZipFile
+
+import pytest
+from django.conf import settings
+from django.db import DatabaseError
+
+from jobserver.releases import Release, handle_release, hash_files
+
+from ..factories import BackendFactory, ReleaseFactory, WorkspaceFactory
+
+
+def raise_error(**kwargs):
+    raise DatabaseError("test error")
+
+
+def make_release_zip(release_path):
+    with ZipFile(release_path, "w") as zf:
+        zf.writestr("file.txt", "test")
+
+    # precalculated hash of above file
+    return "098f6bcd4621d373cade4e832627b4f6"
+
+
+@pytest.mark.django_db
+def test_handle_release_created(monkeypatch, tmp_path):
+    monkeypatch.setattr(settings, "RELEASE_STORAGE", tmp_path / "releases")
+    workspace = WorkspaceFactory()
+    backend = BackendFactory()
+
+    upload = tmp_path / "release.zip"
+    release_hash = make_release_zip(upload)
+
+    release, created = handle_release(
+        workspace, backend, "user", release_hash, upload.open("rb")
+    )
+    assert created
+    assert release.id == release_hash
+    assert release.backend_user == "user"
+    assert release.files == ["file.txt"]
+    assert release.file_path("file.txt").read_text() == "test"
+
+
+@pytest.mark.django_db
+def test_handle_release_already_exists(monkeypatch, tmp_path):
+    monkeypatch.setattr(settings, "RELEASE_STORAGE", tmp_path / "releases")
+
+    upload = tmp_path / "release.zip"
+    release_hash = make_release_zip(upload)
+    release = ReleaseFactory(
+        id=release_hash, files=["file.txt"], backend_user="original"
+    )
+
+    release, created = handle_release(
+        release.workspace, release.backend, "user", release_hash, upload.open("rb")
+    )
+    assert not created
+    assert release.id == release_hash
+    assert release.backend_user == "original"
+    assert release.files == ["file.txt"]
+
+
+@pytest.mark.django_db
+def test_handle_release_bash_hash(monkeypatch, tmp_path):
+    monkeypatch.setattr(settings, "RELEASE_STORAGE", tmp_path / "releases")
+
+    upload = tmp_path / "release.zip"
+    release_hash = make_release_zip(upload)
+    release = ReleaseFactory(
+        id=release_hash, files=["file.txt"], backend_user="original"
+    )
+
+    with pytest.raises(Exception):
+        handle_release(
+            release.workspace, release.backend, "user", "bad hash", upload.open("rb")
+        )
+
+
+@pytest.mark.django_db
+def test_handle_release_db_error(monkeypatch, tmp_path):
+    monkeypatch.setattr(settings, "RELEASE_STORAGE", tmp_path / "releases")
+    monkeypatch.setattr(Release.objects, "create", raise_error)
+
+    upload = tmp_path / "release.zip"
+    release_hash = make_release_zip(upload)
+    workspace = WorkspaceFactory()
+    backend = BackendFactory()
+
+    with pytest.raises(DatabaseError):
+        handle_release(workspace, backend, "user", release_hash, upload.open("rb"))
+
```

Comment:
I can't quite get my head around what this test is confirming, that the database is used inside `handle_release` or?

Something I've been trying to do recently is to add comments for my asserts to express what they're intended to check, might be worth it here.

Reply:
There's some missing asserts that somehow got dropped - it's meant to check the files have been deleted.